### PR TITLE
Serve vllm using subprocess instead of multiprocessing

### PR DESCRIPTION
**Pull request by alimaredia**
**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: src/instructlab/model/backends/backends.py [hunk 1]
```diff
@@ -202,7 +203,8 @@ def ensure_server(
     host="localhost",
     port=8000,
     queue=None,
-) -> Tuple[multiprocessing.Process, str]:
+    server_process_func=None,
```
Comment (alimaredia) [inline src/instructlab/model/backends/backends.py:206]: @n1hility @danmcp FYI I couldn't think of a less invasive way to refactor the code than to not set a type annotation for the `server_process_func`. Can you think of anything cleaner?

File: src/instructlab/model/backends/backends.py [hunk 2]
```diff
@@ -202,7 +203,8 @@ def ensure_server(
...
+) -> Tuple[multiprocessing.Process, subprocess.Popen, str]:
```
Comment (alimaredia) [inline src/instructlab/model/backends/backends.py:207]: @n1hility @danmcp I don't feel great about returning this tuple (one process for each backend) but in line with my comment one line aboe it felt like the least invasive thing I could do.

File: src/instructlab/model/backends/vllm.py [hunk 3]
```diff
@@ -41,154 +17,109 @@ def __init__(
...
             raise exc
 
     def shutdown(self):
         """Shutdown vllm server"""
-        # Needed when the chat is connected to a server and the temporary server is started
+        # Needed when a temporary server is started
         if self.process is not None:
-            self.logger.debug("Terminating temporary server.")
             self.process.terminate()
-            self.process.join(timeout=30)
```
Comment (n1hility) [inline src/instructlab/model/backends/vllm.py:107]: you probably want a process.wait() timeout, and on exception process.kill

File: src/instructlab/model/backends/backends.py (hunk 1 above)
Comment (n1hility) [reply src/instructlab/model/backends/backends.py:228]: we discussed options on chat, decided this cleanup can wait till later

File: src/instructlab/model/backends/backends.py (hunk 2 above)
Comment (n1hility) [reply src/instructlab/model/backends/backends.py:229]: this is good practice IMO could always improve the abstraction later

**Review by n1hility — APPROVED** [review]
My only lingering concern is that the introduction of vllm and llamacpp args contradicts our dev doc choice. Although after discussing with Ali, following up in a later PR is a good idea to unblock eval.

**Review by n1hility — APPROVED** [review]

**Review by danmcp — DISMISSED** [review]

File: src/instructlab/model/backends/backends.py [hunk 4]
```diff
@@ -248,41 +270,71 @@ def ensure_server(
...
         temp_api_base = get_api_base(host_port)
         logger.debug(f"Starting a temporary server at {temp_api_base}...")
+        llama_cpp_server_process = None
+        vllm_server_process = None
+
+        if backend == VLLM:
+            # TODO: resolve how the hostname is getting passed around the class and this function
+            vllm_server_process = server_process_func(port=port)
+            count = 0
+            # TODO should this be configurable?
```
Comment (nathan-weinberg) [inline src/instructlab/model/backends/backends.py:280]: Agree but this can definitely be a followup

**Review by nathan-weinberg — APPROVED** [review]
Just sat with @alimaredia who talked me through all my comments

**Review by n1hility — APPROVED** [review]

**Review by danmcp — APPROVED** [review]

File: src/instructlab/model/chat.py [hunk 5]
```diff
@@ -176,43 +175,10 @@ def chat(
...
-                model_path=model_path,
-                model_family=model_family,
-                host=host,
-                port=port,
-            )
+        ctx.obj.config.serve.llama_cpp.llm_family = model_family
+        backend_instance = backends.select_backend(logger, ctx.obj.config.serve)
 
         try:
             # Run the llama server
```
Comment (danmcp) [inline src/instructlab/model/chat.py:184]: I could be reading this wrong and it was already the case in the code, but is this run connected to the finally below?  And shouldn't it be?

**Pull request merged**